### PR TITLE
Defer pandas import on worker in P2P shuffle

### PR DESCRIPTION
**Pull request by gjoseph92**
In #5688, we discovered that #5520 was increasing worker unmanaged memory usage by ~10MiB at startup (on my mac). I suspect that this came from importing pandas.

The shuffle extension already worked if pandas wasn't available: it just wouldn't install itself on the worker. However, to check if pandas was available, it was importing pandas—so if pandas _was_ available, every worker would have to spend the time and memory to import pandas at startup, even if it wasn't used at all.

With this PR, all use of pandas is deferred. There's also now an explicit test that importing the shuffle does not import pandas.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/shuffle/shuffle.py [hunk 1]
```diff
@@ -53,6 +54,8 @@ def rearrange_by_column_p2p(
     column: str,
     npartitions: int | None = None,
 ):
+    from dask.dataframe import DataFrame
```
Comment (gjoseph92) [inline distributed/shuffle/shuffle.py:57]: I don't love how `dask.dataframe.DataFrame` is also imported `if TYPE_CHECKING`. It means I could remove this line here, and the code would still pass mypy and flake8, even though it would fail at runtime. Tests would still fail, but I could see making a similar mistake in the future in some place that didn't have test coverage, and unwittingly introducing a runtime error.

However, because I want to use `DataFrame` as a type annotation, it does need to be imported for type-checkers.

I don't really think there's a way around this. Just curious if anyone had ideas, particularly @crusaderky or @ian-r-rose

File: distributed/shuffle/tests/test_no_pandas_import.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+
+import subprocess
+import sys
+
+import_check_code = """
```
Comment (gjoseph92) [inline distributed/shuffle/tests/test_no_pandas_import.py:6]: I tried a couple other options to avoid this kind of nasty "code in a string" pattern:
1. Mucking with `sys.modules` (setting `pandas` to None, removing `dask.dataframe` and `distributed.shuffle`, then re-importing `distributed.shuffle`). This worked, but felt brittle, because you basically had to list all the places `pandas` might be transitively imported from, and remove them from `sys.modules`.
3. Using `multiprocessing` in `spawn` mode to start a fresh interpreter and run this code. Because `spawn` mode works by sending pickled function, the the function must be importable. And un-picking the function means importing `distributed.shuffle.tests.test_no_pandas_import.func_name_whatever`. Therefore, `distributed.shuffle` gets imported before your function can even execute! So though you can assert that pandas hasn't been imported, if it has been imported, you can't tell from where, which makes the test very annoying to debug.

In the end, the "code in a string" was both much easier to read and more robust than anything else I could come up with.

(Also, if distributed had a mindeps test, we might not need this at all. But it's nice to have anyway for local development.)

File: distributed/shuffle/shuffle.py (hunk 1 above)
Comment (crusaderky) [reply distributed/shuffle/shuffle.py:57]: Yep, this issue is well-known upstream: https://github.com/PyCQA/pyflakes/pull/530

File: distributed/shuffle/tests/test_no_pandas_import.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+def test_import_no_pandas():
+    p = subprocess.run(
+        [sys.executable],
+        input=import_check_code,
+        text=True,
+        timeout=10,
+    )
+    assert (
+        p.returncode == 0
+    ), "Importing the shuffle extension without pandas failed. See logs for details."
```
Comment (crusaderky) [inline distributed/shuffle/tests/test_no_pandas_import.py:32]:
```suggestion
    subprocess.check_call(
        [sys.executable],
        input=import_check_code,
        text=True,
    )
```
The 10s timeout made me nervous. I would not be surprised if it caused flakiness on our CI.

File: distributed/shuffle/tests/test_no_pandas_import.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
+
+import subprocess
+import sys
+
+import_check_code = """
+import sys
```
Comment (crusaderky) [inline distributed/shuffle/tests/test_no_pandas_import.py:7]:
```suggestion
import sys
from .shuffle import ShuffleWorkerExtension
```
otherwise this test doesn't test anything useful...

File: distributed/shuffle/tests/test_no_pandas_import.py (hunk 3 above)
Comment (crusaderky) [inline distributed/shuffle/tests/test_no_pandas_import.py:32]:
```suggestion
import sys

from distributed.nanny import Nanny
from distributed.utils_test import gen_cluster


@gen_cluster(client=True, Worker=Nanny)
async def test_import_no_pandas(c, s, a, b):
    def assert_no_pandas(dask_worker):
        assert dask_worker.extensions["shuffle"]
        assert "pandas" not in sys.modules
        
    await c.run(assert_no_pandas)
```

File: distributed/shuffle/tests/test_no_pandas_import.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+from __future__ import annotations
```
Comment (jrbourbeau) [inline distributed/shuffle/tests/test_no_pandas_import.py:1]: Thanks @gjoseph92. Can we move this into `distributed/shuffle/tests/test_shuffle_extension.py`? It seems like it doesn't merit it's own test module

File: distributed/shuffle/tests/test_no_pandas_import.py (hunk 3 above)
Comment (gjoseph92) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: This is a neat way of doing it. The one thing is that it won't point you to _where_ pandas is being imported, just that it is (somewhere). The reason I went for running the script in a subprocess was that you'd get an import error (and traceback to) exactly where the import was happening, which made fixing the failure much easier.

If you'd prefer giving up the debuggability for this more conventional pattern, I'm happy to make the change.

File: distributed/shuffle/tests/test_no_pandas_import.py (hunk 5 above)
Comment (gjoseph92) [reply distributed/shuffle/tests/test_no_pandas_import.py:1]: I separated it because the subprocess pattern felt a bit distinct from those unit tests, but sure.

File: distributed/shuffle/tests/test_no_pandas_import.py (hunk 3 above)
Comment (crusaderky) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: I think that git bisect is enough to offer debuggability?

Comment (gjoseph92) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: Even still, you have to figure out what line of code is causing the import (and know how to use git bisect). Getting the traceback to the exact line of code just feels more friendly to other contributors.

Comment (crusaderky) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: I counter-argue that your version would not notice it if there was an ``import pandas`` inside ``ShuffleWorkerExtension.__init__``. IMHO tests should be as black-box as possible.

Comment (gjoseph92) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: That's a good point.

Unfortunately, the `gen_cluster` approach doesn't work. Pandas is already imported even if the shuffle extension is not imported on the worker. A good example of how when there isn't a traceback to see where the import is happening, debugging it is tough!

My guess is that the problem is that nannies pre-import pandas (and other modules) automatically into the worker process:
The global `mp_context` used by Nannies to create worker processes is initialized here at import time: https://github.com/dask/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/utils.py#L69-L93
In particular, notice how all required and optional packages that are importable are added to `set_forkserver_preload`.

pandas is in the list of optional packages: https://github.com/dask/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/versions.py#L24-L29

(Sidenote: we're using [`spawn` by default](https://github.com/gjoseph92/distributed/blob/338d0be2e50b5458ef019ec8ce7377bda3097ef1/distributed/distributed.yaml#L77); I confirmed that's what's used when I run the test. Reading the [`multiprocessing` code](https://github.com/python/cpython/blob/dad6be5ffe48beb74fad78cf758b886afddc7aed/Lib/multiprocessing/context.py#L180-L185), I don't understand why `set_forkserver_preload` is affecting non-forkserver contexts. But if I comment out the pandas line from `optional_packages`, the test passes, so I'm not going to look into it much further.)

Comment (gjoseph92) [reply distributed/shuffle/tests/test_no_pandas_import.py:32]: For posterity: @crusaderky and I discovered that `set_forkserver_preload` has no effect when using `spawn`, as expected. Instead, the reason pandas was always imported on Nanny workers (but not imported if it was commented out from the list in `versions.py`) is that the `get_versions` check on client connect imports all those packages listed in order to get their versions. This could be avoided.

File: distributed/tests/test_nanny.py [hunk 6]
```diff
@@ -610,3 +611,21 @@ async def test_environ_plugin(c, s, a, b):
         assert results[a.worker_address] == "123"
         assert results[b.worker_address] == "123"
         assert results[n.worker_address] == "123"
+
+
+@pytest.mark.xfail(reason="https://github.com/dask/distributed/issues/5723")
+@gen_cluster(client=True, Worker=Nanny, nthreads=[("", 1)])
+async def test_no_unnecessary_imports_on_worker(c, s, a):
```
Comment (crusaderky) [inline distributed/tests/test_nanny.py:618]:
```suggestion
@pytest.mark.parametrize("modname", ["numpy", "pandas"])
@gen_cluster(client=True, Worker=Nanny, nthreads=[("", 1)])
async def test_no_unnecessary_imports_on_worker(c, s, a, modname):
```

File: distributed/tests/test_nanny.py [hunk 7]
```diff
@@ -610,3 +611,21 @@ async def test_environ_plugin(c, s, a, b):
...
+    Importing modules like pandas slows down worker startup, especially if workers are
+    loading their software environment from NFS or other non-local filesystems.
+    It also slightly increases memory footprint.
+    """
+
+    def assert_no_pandas(dask_worker):
+        assert "pandas" not in sys.modules
+
+    await c.wait_for_workers(1)
+    await c.run(assert_no_pandas)
```
Comment (crusaderky) [inline distributed/tests/test_nanny.py:631]:
```suggestion
    out = await c.run(lambda: modname in sys.modules)
    assert out[a.address] is False
```

Comment (crusaderky) [inline distributed/tests/test_nanny.py:631]:
```suggestion
    def assert_no_import(dask_worker):
        assert modname not in sys.modules

    await c.wait_for_workers(1)
    await c.run(assert_no_import)
```

**Comment by crusaderky** [comment]
...Great, now test_memory is heavily flaky -__-

File: distributed/tests/test_nanny.py [hunk 8]
```diff
@@ -610,3 +611,22 @@ async def test_environ_plugin(c, s, a, b):
         assert results[a.worker_address] == "123"
         assert results[b.worker_address] == "123"
         assert results[n.worker_address] == "123"
+
+
+@pytest.mark.xfail(reason="https://github.com/dask/distributed/issues/5723")
+@pytest.mark.parametrize("modname", ["numpy", "pandas"])
```
Comment (crusaderky) [inline distributed/tests/test_nanny.py:617]:
```suggestion
@pytest.mark.parametrize(
    "modname",
    [
        pytest.param(
            "numpy",
            marks=pytest.mark.xfail(reason="distributed#5723, distributed#5729"),
        ),
        "scipy",
        pytest.param("pandas", marks=pytest.mark.xfail(reason="distributed#5723")),
    ],
)
```

File: distributed/tests/test_nanny.py (hunk 6 above)
Comment (gjoseph92) [reply distributed/tests/test_nanny.py]: NumPy currently seems to always be imported (AFAICT locally). I dropped a breakpoint in `numpy/__init__.py` and got:
```
-> import distributed
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/__init__.py(7)<module>()
-> from .actor import Actor, ActorFuture
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/actor.py(5)<module>()
-> from .client import Future
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/client.py(59)<module>()
-> from .batched import BatchedSend
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/batched.py(10)<module>()
-> from .core import CommClosedError
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/core.py(27)<module>()
-> from . import profile, protocol
  <frozen importlib._bootstrap>(1058)_handle_fromlist()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/protocol/__init__.py(4)<module>()
-> from .compression import compressions, default_compression
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/distributed/distributed/protocol/compression.py(20)<module>()
-> import blosc
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/dask-playground/env/lib/python3.9/site-packages/blosc/__init__.py(85)<module>()
-> from blosc.test import run as test
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
  /Users/gabe/dev/dask-playground/env/lib/python3.9/site-packages/blosc/test.py(15)<module>()
-> import numpy
  <frozen importlib._bootstrap>(1007)_find_and_load()
  <frozen importlib._bootstrap>(986)_find_and_load_unlocked()
  <frozen importlib._bootstrap>(680)_load_unlocked()
  <frozen importlib._bootstrap_external>(790)exec_module()
  <frozen importlib._bootstrap>(228)_call_with_frames_removed()
> /Users/gabe/dev/dask-playground/env/lib/python3.9/site-packages/numpy/__init__.py(110)<module>()
```

tl;dr: compression tries to import blosc, blosc imports its own testing suite, that test suite imports numpy.

If I remove blosc (`sys.modules["blosc"] = None`), and use this PR plus https://github.com/dask/distributed/pull/5724, I can start a Nanny worker without import NumPy.

**Comment by crusaderky** [comment]
Merged CI test of this #5695 + #5724 running on https://github.com/crusaderky/distributed/runs/4983212568 and things don't look very good so far. I can see much more flakiness than main and much slower end-to-end test runtime.

**Comment by crusaderky** [comment]
It seems like CI is back on track. I reran all unit tests.
``test_memory`` is heavily flaky; I can also see a failure in ``distributed/tests/test_active_memory_manager.py::test_drop_with_paused_workers_with_running_tasks_5`` which may or may not be related. I'm working on them.

**Comment by crusaderky** [comment]
Blocked by
-  https://github.com/dask/distributed/pull/5737

**Comment by crusaderky** [comment]
@gjoseph92 please merge from main; you shoud get all green now (well, no more red than usual at least :laughing: )

**Comment by gjoseph92** [comment]
Thanks @crusaderky for all your work smoothing out the flakes!

**Comment by gjoseph92** [comment]
```
FAILED distributed/tests/test_worker.py::test_worker_reconnects_mid_compute
FAILED distributed/tests/test_worker.py::test_worker_reconnects_mid_compute_multiple_states_on_scheduler
```
Just on one windows box! Looking very good.

**Comment by crusaderky** [comment]
Thank you!

**Comment by rjzamora** [comment]
@crusaderky @gjoseph92 - Just a note that this PR seems to have broken Dask's [test_scheduler_highlevel_graph_unpack_import](https://github.com/dask/dask/blob/eda286bbe68e11334e4784097210fe09f255b3d9/dask/tests/test_layers.py#L201) tests.

It doesn't look like this PR does anything dangerous, since pandas does not **need** to be available. However, the Dask tests currently assume that the scheduler will not import pandas/numpy.  cc @jrbourbeau - In case you have thoughts on `test_scheduler_highlevel_graph_unpack_import`?
